### PR TITLE
Logging Architecture

### LINKED ISSUE #320: Create architecture diagrams and docs for Loki Logs on new Observability cluster

### PR DESCRIPTION
**Pull request by bnshr**
Describe the NERC logging architecture by providing logical and schematic digram.
fixes https://github.com/nerc-project/operations/issues/320

### CODE REVIEW

File: architecture/logging/portfolio.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Logging System
+
+For better understanding of the development, test, deploy and
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
For better understanding of the development, test, deploy, and operation, the following diagram identified several parts of the system that composed the structure of the logging system in NERC.
```

File: architecture/logging/portfolio.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+operation, the following diagram identified several parts of the
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
```

File: architecture/logging/portfolio.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+system that composed the structure of the logging system in NERC.
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
```

File: architecture/logging/portfolio.md (hunk 1 above)
Comment (schwesig) [reply architecture/logging/portfolio.md]: - avoiding to split one sentence over multiple line

File: architecture/logging/portfolio.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+Logs are written by the clusters managed by Advanced Cluster Management (ACM) and read by the Cluster admins with read permission only.
+
+The below digram depicts the read and write flow. However, please bear in mind that only essential components are identified for the sake of brevity as lots are happening behind the scene.
+
+![plot](./images/Logging_Schematic_Diagram.jpg)
+
+
+## Read path
+
+1. The read request comes from the admin users who have the permission
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
1. The read request comes from the admin users who have the permission (by adding into [nerc-logs-metrics](https://github.com/orgs/OCP-on-NERC/teams/nerc-logs-metrics) group in GitHub) to read the logs from the OpenShift Production (as an example) Cluster Console.
```

File: architecture/logging/portfolio.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+(by adding into nerc-logs group in GitHub) to read the logs from
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
```

File: architecture/logging/portfolio.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+the OpenShift Production (as an example) Cluster Console.
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
```

File: architecture/logging/portfolio.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+This is enabled for them by enabling logging Console Plugin after installing
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
This is enabled for them by enabling logging Console Plugin after installing OpenShift Logging Operator.
```

File: architecture/logging/portfolio.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+OpenShift Logging Operator.
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
```

File: architecture/logging/portfolio.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+2. The query request is sent to the LokiStack in the Observability cluster and is handled by the Querier service. This service processes queries using LogQL query language, fetching logs both from the ingesters and from the long-term storage.
+
+3. To fetch the log data from the long-term storage, the request is forwarded to the NooBaa operator, one of the operators in the OpenShift Data Foundation installed in Observability cluster. This operator is responsible to manage Multicloud Object Gateway components, such as object bucket claim, bucket classes and backing store in the OpenShift Data Foundation.
+
+4. The backing store handles the read and write request from the customer, as it provides MCG
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
4. The backing store handles the read and write request from the customer, as it provides MCG the ability to save the data of the provisioned bucket on top of it. In this architecture, the backing store is AWS:S3 bucket.
```

File: architecture/logging/portfolio.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+the ability to save the data of the provisioned bucket on top of it. In this architecture, the backing store is AWS:S3 bucket.
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
```

File: architecture/logging/portfolio.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+For better understanding of the development, test, deploy and
+operation, the following diagram identified several parts of the
+system that composed the structure of the logging system in NERC.
+
+
+![plot](./images/Logging_Logical_Diagram.jpg)
+
+
+## Schematic Digram
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
## Schematic Diagram
```

File: architecture/logging/portfolio.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+Logs are written by the clusters managed by Advanced Cluster Management (ACM) and read by the Cluster admins with read permission only.
+
+The below digram depicts the read and write flow. However, please bear in mind that only essential components are identified for the sake of brevity as lots are happening behind the scene.
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
The below diagram depicts the read and write flow. However, please bear in mind that only essential components are identified for the sake of brevity as lots are happening behind the scene.
```

File: architecture/logging/portfolio.md [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+1. The read request comes from the admin users who have the permission
+(by adding into nerc-logs group in GitHub) to read the logs from
+the OpenShift Production (as an example) Cluster Console.
+This is enabled for them by enabling logging Console Plugin after installing
+OpenShift Logging Operator.
+
+2. The query request is sent to the LokiStack in the Observability cluster and is handled by the Querier service. This service processes queries using LogQL query language, fetching logs both from the ingesters and from the long-term storage.
+
+3. To fetch the log data from the long-term storage, the request is forwarded to the NooBaa operator, one of the operators in the OpenShift Data Foundation installed in Observability cluster. This operator is responsible to manage Multicloud Object Gateway components, such as object bucket claim, bucket classes and backing store in the OpenShift Data Foundation.
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
3. To fetch the log data from the long-term storage, the request is forwarded to the NooBaa operator, one of the operators in the OpenShift Data Foundation installed in Observability cluster. This operator is responsible to manage Multicloud Object Gateway components, such as object bucket claim, bucket classes, and backing store in the OpenShift Data Foundation.
```

File: architecture/logging/portfolio.md [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+4. The backing store handles the read and write request from the customer, as it provides MCG
+the ability to save the data of the provisioned bucket on top of it. In this architecture, the backing store is AWS:S3 bucket.
+
+6. A S3 bucket is provisioned by a Object Bucket Claim for storing data. And the route for the S3 access is provided by the interal CRD called NooBaa.
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
6. A S3 bucket is provisioned by a Object Bucket Claim for storing data. And the route for the S3 access is provided by the internal CRD called NooBaa.
```

File: architecture/logging/portfolio.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+> NERC logs are stored inside Log Bucket. Since the logs are purged to keep log size under control, there is also backup of the logs and its corresponding ObjectBucketClaim. The read and write request of logs only involve this Log bucket from any ACM cluster, however Observability cluster can make read request to the Log Backup bucket as well. <br/><br/>The data sync between these two buckets is provisioned by mirroring feature in OpenShift Data Foundation.
+
+10. After getting access to the logs from the bucket, the query returns the data following the reverse path. Query result is sent to the Query frontend instead of Querier to display to the admin users who requested for the log data.
+
+> The LOKI ruler continuously evaulates a set of configurable queries and performs an action based on the result. This component is not directly involved during the read request from the cluster admin users, however to generate alerts it reads the logs from the log store following the same path in the OpenShift Data Foundation.
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
> The LOKI ruler continuously evaluates a set of configurable queries and performs an action based on the result. This component is not directly involved during the read request from the cluster admin users, however to generate alerts, it reads the logs from the log store following the same path in the OpenShift Data Foundation. 
```

File: architecture/logging/portfolio.md [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+
+10. After getting access to the logs from the bucket, the query returns the data following the reverse path. Query result is sent to the Query frontend instead of Querier to display to the admin users who requested for the log data.
+
+> The LOKI ruler continuously evaulates a set of configurable queries and performs an action based on the result. This component is not directly involved during the read request from the cluster admin users, however to generate alerts it reads the logs from the log store following the same path in the OpenShift Data Foundation.  
+
+
+## Write path
+
+To aggegrate the logs (application, infrastructure and audit) the cluster logging operator is installed in the ACM production clusters. This operator components collects and forwards logs to the log store to be saved for reading purpose later.
```
Comment (schwesig) [inline architecture/logging/portfolio.md]:
```suggestion
To aggegrate the logs (application, infrastructure, and audit) the cluster logging operator is installed in the ACM production clusters. This operator components collects and forwards logs to the log store to be saved for reading purpose later.
```

**Review by schwesig — CHANGES_REQUESTED** [review]
just a few minor changes (the team name and some typos)

File: architecture/logging/portfolio.md (hunk 4 above)
Comment (schwesig) [reply architecture/logging/portfolio.md]: - avoid multiple lines
- changing team nerc-logs to nerc-logs-metrics and adding link

**Review by computate — APPROVED** [review]
Looks like the checks are passing now, nice work!
I love all the relevant links in the document, it's very good.
The diagrams are beautiful and clear.

**Review by schwesig — APPROVED** [review]
awesome job @bnshr
approved!

**Comment by bnshr** [comment]
Thanks for the review and suggestions. 👍

**Pull request merged**